rfcs/0010-stamps.md

```diff
@@ -0,0 +1,37 @@
+# Stamps: Airdrops with Intent
+
+The goal of this RFC is to define an airdrop mechanism that authorizes funds
+for a particular use.
+
+## Background
+
+To run an on-chain program, a client must first acquire tokens to pay a Solana
+cluster a transaction fee to instantiate the program. To acquire the tokens,
+the client might request them from a special client called a *drone*.
+
+## Problems
+
+1. Once the client has tokens to instantiate the program, the client may choose
+   to simply keep them.
+2. Loading a program requires two transaction fees - one for the airdrop and
+   one to load the program.
+
+## Stamps
+
+Instead of a drone sending a payment transaction to the cluster, it could send
+a signed *load* transaction back to the client. This signed transaction would be
+called a stamp. Like how a food stamp may only be used to purchase food, a
+drone's load stamp may only be used to load a program. It doesn't matter to the
+cluster whether it receives a stamp from the client or the drone, only that the
+signature is valid.
+
+## Example usage
+
+Creator of on-chain game tic-tac-toe hosts a drone that responds to airdrop
```

Comment:
BYO**D** seem tedious and very anti-chain.  Perhaps an account can signal somehow that it's willing to pay the transaction fee for certain kinds of instructions, so joe public can get bootstrapped without using the s word (server that is).

Reply:
I agree. The general theme of Solana has been to move any off-chain functionality onto the chain.  I see this stamps technique as sufficient, not ideal.  In the current system, the fee is collected before the instruction is executed. We'd need to add another "who's buying?" entrypoint.  A separate RFC? Perhaps even described as one intending to render this one obsolete?

Reply:
If stamps are not ideal, what is driving this RFC?  Can we wait until we can implement the ideal?

Reply:
What's driving the RFC is that the Tabletops release should require fees for all transactions but wanting to continue offering on-chain games that can be played by users without any tokens. This RFC offers a convention for doing that without any rearchitecture.

Reply:
@aeyakovenko, @mvines wants a drone-less solution.

Reply:
@mvines who decides the "are you a bot" question?

Reply:
@garious @mvines I am describing a dynamic server that doesn't need to "airdrop".  Server signs a tx with `key[0]`
```
signed_key[0] - drone key, pay the fee
signed_key[1] - player key
key[0] - game key
key[1] - player key
instruction - { signed_keys: 1, keys: 0,1}
```
so the airdop is gone

Reply:
Are you paraphrasing this RFC?

Reply:
I don't see the point of the user signature. In this example, they are being given a game instance.

Reply:
@garious upon closer inspection, yes :). I think the "stamp" wording is confusing.  The functional part is multiple signatures per transaction, and instructions referencing any subset of the transaction keys.  It's hard to infer that from the rfc

Reply:
> I don't see the point of the user signature. In this example, they are being given a game instance.

@garious if the original program requires the signature for the instruction's 0 key, the user would supply it.  It would be up to the program.

Reply:
@aeyakovenko, I don't see what problem is solved with a user signature. Maybe there's something I can do to clarify the scope of the RFC?